notebooks/week2/05.log_and_register_fe_model.py

```diff
@@ -0,0 +1,182 @@
+schema_name = config.schema_name
+mlflow_experiment_name = config.mlflow_experiment_name
+
+# Define table names and function name
+feature_table_name = f"{catalog_name}.{schema_name}.power_features"
+function_name = f"{catalog_name}.{schema_name}.temperature_rounded"
+
+
+# COMMAND ----------
+# Load training and test sets
+train_set = spark.table(f"{catalog_name}.{schema_name}.train_set_nico")
+test_set = spark.table(f"{catalog_name}.{schema_name}.test_set_nico")
+
+
+# COMMAND ----------
+# Create or replace the power_features table
+spark.sql(f"""
+CREATE OR REPLACE TABLE {catalog_name}.{schema_name}.power_features
+(DateTime STRING NOT NULL,
+ Temperature DOUBLE,
+ Humidity DOUBLE,
+ Wind_Speed DOUBLE);
+""")
+
+spark.sql(f"ALTER TABLE {catalog_name}.{schema_name}.power_features "
+          "ADD CONSTRAINT power_pk PRIMARY KEY(DateTime);")
+
+spark.sql(f"ALTER TABLE {catalog_name}.{schema_name}.power_features "
+          "SET TBLPROPERTIES (delta.enableChangeDataFeed = true);")
+
+# Insert data into the feature table from both train and test sets
+spark.sql(f"INSERT INTO {catalog_name}.{schema_name}.power_features "
+          f"SELECT DateTime, Temperature, Humidity, Wind_Speed FROM {catalog_name}.{schema_name}.train_set_nico")
+spark.sql(f"INSERT INTO {catalog_name}.{schema_name}.power_features "
+          f"SELECT DateTime, Temperature, Humidity, Wind_Speed FROM {catalog_name}.{schema_name}.test_set_nico")
+
+# COMMAND ----------
+# Define a function to calculate the power's age using the current year and YearBuilt
+spark.sql(f"""
+CREATE OR REPLACE FUNCTION {function_name}(Temperature DOUBLE)
+RETURNS INT
+LANGUAGE PYTHON AS
+$$
+return round(Temperature)
+$$
+""")
+# COMMAND ----------
+# Load training and test sets
+train_set = spark.table(f"{catalog_name}.{schema_name}.train_set_nico").drop("Temperature", "Humidity", "Wind_Speed")
+test_set = spark.table(f"{catalog_name}.{schema_name}.test_set_nico").toPandas()
+
+# Cast YearBuilt to int for the function input
+# train_set = train_set.withColumn("RoundedTemp", train_set["RoundedTemp"].cast("int"))
+# train_set = train_set.withColumn("DateTime", train_set["DateTime"].cast("string"))
+# COMMAND ----------
+
+# TODO: This leads to unauthorized error as Maria said
+
+# Feature engineering setup
+
+training_set = fe.create_training_set(
+    df=train_set,
+    label=target,
+    feature_lookups=[
+        FeatureLookup(
+            table_name=feature_table_name,
+            feature_names=["Temperature", "Humidity", "Wind_Speed"],
+            lookup_key="DateTime",
+        ),
+        FeatureFunction(
+            udf_name=function_name,
+            output_name="temperature_rounded",
+            input_bindings={"Temperature": "RoundedTemp"},
+        ),
+    ],
+    # exclude_columns=["bla"]
+)
+# COMMAND ----------
+
+# Load feature-engineered DataFrame
+training_df = training_set.load_df().toPandas()
+
+# Split features and target
+X_train = training_df[num_features]
```

Comment:
I noticed, temperature_rounded (from Feature Function) is not part of num_features, so will not be used for training.

Reply:
fixed here https://github.com/end-to-end-mlops-databricks/marvelous-databricks-course-NiGroUni/commit/b9c253b44dc5547adbc0e25094cb38f7c807fd2f